### PR TITLE
REVSDL-2665 Change logic for updated protocol of Climate control

### PR DESCRIPTION
**Pull request by VladSemenyuk**

### CODE REVIEW

File: src/components/can_cooperation/include/can_cooperation/can_module_constants.h [hunk 1]
```diff
@@ -239,14 +239,21 @@ const char kState[]                 = "state";
...
 const char kAutoModeEnable[]     = "autoModeEnable";
 const char kDefrostZone[]        = "defrostZone";
 const char kDualModeEnable[]     = "dualModeEnable";
+const char kVentilationMode[] = "ventilationMode";
 // ClimateControlData struct
 
+// Temperature struct
+const char kUnit[] = "unit";
+const char kValueC[] = "valueC";
+const char kValueF[] = "valueF";
```
Comment (bonewell) [inline src/components/can_cooperation/include/can_cooperation/can_module_constants.h]: Bad names, isn't it?

Comment (VladSemenyuk) [inline src/components/can_cooperation/include/can_cooperation/can_module_constants.h]: I didn't invent them. Just call variables like field name in xml. It is common approach. Цhen I look at the variable name, I immediately see the field name in a json.

**Comment by VladSemenyuk** [comment]
@DKlimenko

File: src/components/can_cooperation/include/can_cooperation/can_module_constants.h (hunk 1 above)
Comment (bonewell) [inline src/components/can_cooperation/include/can_cooperation/can_module_constants.h]: Yeah, I understand your point. I say about our protocol :)

Comment (VladSemenyuk) [inline src/components/can_cooperation/include/can_cooperation/can_module_constants.h]: It is not protocol review. ;))))

File: src/components/can_cooperation/src/validators/struct_validators/climate_control_data_validator.cc [hunk 2]
```diff
@@ -125,8 +110,32 @@ ValidationResult ClimateControlDataValidator::Validate(const Json::Value& json,
 
   ValidationResult result = ValidateSimpleValues(json, outgoing_json);
 
+  if (result != ValidationResult::SUCCESS) {
+    return result;
+  }
+
+  if (IsMember(json, kCurrentTemp)) {
+    result = ClimateControlDataValidator::instance()->Validate(
```
Comment (bonewell) [inline src/components/can_cooperation/src/validators/struct_validators/climate_control_data_validator.cc]: I didn't catch why you invoke to "this" using singleton?

File: src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  value_f_[ValidationParams::MAX_VALUE] = 90;
+  value_f_[ValidationParams::ARRAY] = 0;
+  value_f_[ValidationParams::MANDATORY] = 0;
+
+  validation_scope_map_[kValueF] = &value_f_;
+  validation_scope_map_[kValueC] = &value_c_;
+}
+
+ValidationResult TemperatureValidator::Validate(const Json::Value& json,
+                                                 Json::Value& outgoing_json) {
```
Comment (bonewell) [inline src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc]: Wrong indent.

File: src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  if (json.isMember(kUnit)) {
+    if (enums_value::kFahrenheit == json[kUnit]) {
+      if (!json.isMember(kValueF)) {
+        LOG4CXX_ERROR(logger_, "Param " <<kValueF <<"is missed!");
```
Comment (bonewell) [inline src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc]: Fix spaces please.

File: src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        return INVALID_DATA;
+      }
+
+      json.removeMember(kValueC);
+    } else if (enums_value::kCelsius == json[kUnit]) {
+      if (!json.isMember(kValueC)) {
+        LOG4CXX_ERROR(logger_, "Param " <<kValueC <<"is missed!");
```
Comment (bonewell) [inline src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc]: Fix spaces, please.

File: src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        return INVALID_DATA;
+      }
+
+      json.removeMember(kValueF);
+    } else {
+      LOG4CXX_ERROR(logger_, "Wrong TemperatureUnit enum value!");
+      return ValidationResult::INVALID_DATA;
+    }
+  } else {
+    LOG4CXX_ERROR(logger_, "Mandatory param " <<kUnit <<"is missed!");
```
Comment (bonewell) [inline src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc]: Fix spaces, please.

File: src/components/can_cooperation/src/validators/validator.cc [hunk 7]
```diff
@@ -136,7 +136,7 @@ ValidationResult Validator::Validate(const Json::Value& json,
 
     return ValidateIntValue(json.asInt(), validation_scope);
   } else if (ValueType::DOUBLE == validation_scope[ValidationParams::TYPE]) {
-      if (!json.isDouble()) {
+      if ((!json.isDouble()) && (!json.isInt())) {
```
Comment (bonewell) [inline src/components/can_cooperation/src/validators/validator.cc]: Strange, extremely strange. Could you explain?

File: src/components/can_cooperation/src/validators/struct_validators/climate_control_data_validator.cc (hunk 2 above)
Comment (bonewell) [inline src/components/can_cooperation/src/validators/struct_validators/climate_control_data_validator.cc]: Perhaps, you want to use TemperatureValidator, don't you?

Comment (VladSemenyuk) [inline src/components/can_cooperation/src/validators/struct_validators/climate_control_data_validator.cc]: Yes, of course.

File: src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc (hunk 4 above)
Comment (VladSemenyuk) [inline src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc]: There are no problem with spaces.

File: src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc (hunk 5 above)
Comment (VladSemenyuk) [inline src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc]: There are no problem with spaces. CTRL+SHIFT+F used.

Comment (VladSemenyuk) [inline src/components/can_cooperation/src/validators/struct_validators/temperature_validator.cc]: You mean in log text, i understand

**Comment by DKlimenko** [comment]
Reviewed

**Pull request merged**